src/Npgsql/VolatileResourceManager.cs

```diff
@@ -326,6 +375,8 @@ void Dispose()
                 else
                     _connector.Close();
             }
+            _isRMCall.Dispose();
+            _secondPhaseLock.Dispose();
```

Comment:
`Dispose` should be used only if all operation on the semaphore have completed. We have un-referenced the RM for avoiding new waiting threads, then released the lock. But does it immediately cause the waiting threads to no more use the semaphore? I am a bit unsure here, so I have put that try catch around the wait on the semaphore, in case there is a race condition between the signalling with `Release`, the actual end of thread waits, and the disposal here.